### PR TITLE
Maxpool and softmax

### PR DESCRIPTION
**Pull request by vb000**
- `MaxPool2d` and `LogSoftmax` kernels.
- Moved convolution offloading to `Convolution.cpp`
- Fixed dispatch warnings.
- Filter out `-nostdlib` flag when compiling kernels for cosim -- to use C math library.

### CODE REVIEW

**Comment by vb000** [comment]
All pass on cosim.
```
============ 88 passed, 17 skipped, 2 xfailed in 697.64s (0:11:37) =============
BSG REGRESSION TEST PASSED
BSG COSIM PASS: Test passed!
$finish called from file "/mnt/users/ssd1/no_backup/bandhav/bsg_bladerunner/bsg_replicant/testbenches/cosim_wrapper.sv", line 57.
$finish at simulation time         191369340000
           V C S   S i m u l a t i o n   R e p o r t 
Time: 191369340000 ps
CPU Time:    701.250 seconds;       Data structure size:   3.6Mb
Mon Mar 23 15:08:15 2020
```

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -1586,6 +1586,7 @@
   dispatch:
     CPU: log_softmax_cpu
     CUDA: log_softmax_cuda
+    HammerBlade: log_softmax_hb
```
Comment (yodada) [inline aten/src/ATen/native/native_functions.yaml:1589]: Exactly what I need! Sweet!

File: aten/src/ATen/Dispatch.h [hunk 2]
```diff
@@ -8,7 +8,7 @@
 
 #define AT_PRIVATE_CASE_TYPE(enum_type, type, ...) \
   case enum_type: {                                \
-    using scalar_t = type;                         \
+    using scalar_t C10_UNUSED = type;              \
```
Comment (yodada) [inline aten/src/ATen/Dispatch.h:11]: What is the usage of this? I think I'm missing sth

File: hammerblade/torch/kernel/kernel_pooling.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+          bsg_tensor_t* output,
+          bsg_tensor_t* input,
+          bsg_tensor_t* indices,
+          int* kH, int* kW,
+          int* sH, int* sW,
+          int* padH, int* padW,
+          int* dilationH, int* dilationW) {
+    auto y = BSGTensor<float>(output);
+    auto x = BSGTensor<float>(input);
+    auto ind = BSGTensor<long long>(indices);
```
Comment (yodada) [inline hammerblade/torch/kernel/kernel_pooling.cpp]: so we can do 64bit arithmetic on HB?

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (vb000) [reply aten/src/ATen/native/native_functions.yaml:1589]: Yep, almost every modern classification network uses this as the last layer.

File: aten/src/ATen/Dispatch.h (hunk 2 above)
Comment (vb000) [reply aten/src/ATen/Dispatch.h:11]: Dispatch macro sets `scalar_t` with `using scalar_t C10_UNUSED = type; `. This may or may not be used in the offloading function. This results in compiler warning about unused typedef. Adding `C10_UNUSED` adds a compile-time attribute to ignore that warning.

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (yodada) [reply aten/src/ATen/native/native_functions.yaml:1589]: Have you looked at any loss function yet? I'm trying to get `CrossEntropyLoss` to work. It somehow uses this logSoftMax kernel

File: hammerblade/torch/kernel/kernel_pooling.cpp (hunk 3 above)
Comment (vb000) [reply hammerblade/torch/kernel/kernel_pooling.cpp]: `int64_t` to `int32_t` could be a destructive cast. So I was hesitant to go ahead and cast it. But in the hindsight, that seems like the right thing to do.

File: aten/src/ATen/Dispatch.h (hunk 2 above)
Comment (yodada) [reply aten/src/ATen/Dispatch.h:11]: Ok! Understand!

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (vb000) [reply aten/src/ATen/native/native_functions.yaml:1589]: No, haven't looked at loss functions yet. Those are definitely needed for training!

File: hammerblade/torch/kernel/kernel_pooling.cpp (hunk 3 above)
Comment (yodada) [reply hammerblade/torch/kernel/kernel_pooling.cpp]: Maybe make this change and otherwise I think this PR is in good shape!

**Review by yodada — APPROVED** [review]
Looks good! Feel free to merge after getting rid of that `long long`! ;)
